### PR TITLE
[BugFix][SFA] Fix GLM5Next sparse MLA sinks and mask/window parameters

### PR DESCRIPTION
**Pull request by zzzzzz198**
### What this PR does / why we need it?
GLM5Next's NoPE sparse MLA path on Ascend A5 traps inside `SparseFlashMla` under sustained load with data parallelism and ACL Graph enabled:

```
error code = 341
"The address for VEC to access UB is out of bounds"
```
The trap is probabilistic and only shows up after roughly a thousand requests at concurrency 128, which made it hard to attribute: AI Core errors are reported asynchronously, so the exception surfaces at an unrelated `synchronize()`, and the plog fills with `error code = 0` / "timeout or trap error" lines from the cores that were merely caught in the fallout.

This PR fixes four defects on that path. Each is described with the operator source that governs it, from `cann/ops-transformer@master`.

#### 1. `sinks` must be owned by the caller

`SparseFlashMla` requires a per-head float32 `sinks` tensor (`op_host/checkers/sinks_checker.cpp:38`). Allocating one inside the adapter would leave it referenced only by the call's kwargs, so ACL Graph capture bakes in an address the allocator reuses as soon as the call returns, and replaying that graph reads freed memory. `SparseMLAMetadataState` now allocates it once, next to its other persistent operator buffers, and the adapter rejects calls that omit it. Also aligned with the operator's parameter checks: `cmp_mask_mode` must be 0 while `cmp_kv` is absent, and `ori_win_left` / `ori_win_right` must be -1 unless `ori_mask_mode` is 4 (`op_host/checkers/mask_checker.cpp:47,57`). A size check was added before the generated plan is copied into its fixed `[1024]` buffer, matching the operator's contract.

#### 2. `ori_mask_mode` must be 0 in the sparse ori_kv scenario

This is the one that causes the trap.

On A5 the operator picks its template from which tensors are present, not from the mask mode, so passing `ori_sparse_indices` together with `ori_mask_mode=3` is accepted and still lands on the sparse template. What the mask mode changes is whether the kernel initialises its softmax accumulators (`op_kernel/arch35/sparse_flash_mla_swa_kernel_arch35.h:241`):

```cpp
constInfo.needInit = 0;
if (constInfo.oriMaskMode != 0) {
    for (bIdx...) {
        if (s1Size > s2Size || s1Size < expectQs) { constInfo.needInit = 1; break; }
    }
} else {
    constInfo.needInit = 1;   // mask mode 0 always initialises
}
```

With mode 3 the condition is only met when some sequence has a query longer than its KV, which never happens on an ordinary decode step, so `needInit` stays 0 and the initialisation is skipped. A sparse gather does not write every accumulator slot, so the skipped slots keep whatever the previous step left in them; the vector unit then computes a UB offset from that residue and walks out of bounds.

Mode 0 is also what the operator expects semantically: the selected indices are themselves the mask, since the indexer only offers causally valid tokens. The A2/A3 path enforces this explicitly (`op_host/sparse_flash_mla_tiling.cpp:542`), and the metadata interface ties the two together from the other side - with `ori_mask_mode=0` and a non-zero `ori_topk` it requires `ori_topk_length` (`sparse_flash_mla_metadata/op_host/sparse_flash_mla_metadata_check.h:447`).

#### 3. The plan and the operator call described different work

`SparseFlashMlaMetadata` produces the core-split plan, and the kernel walks the query rows according to it. The plan was generated during metadata construction, from the padded `cu_seqlens_q` and the full top-k length tensor, while eager and piecewise steps trim the query to the unpadded token count before calling the operator. Observed in practice:

```
query has 1 rows but ori_topk_length was planned over 418 rows (num_actual_tokens=1, num_input_tokens=418, cudagraph_mode=NONE)
```

The gap comes from data-parallel padding: a rank with one real token still runs the group-wide token count. The operator's TND contract requires the first dimension of `q` to equal the last element of `actual_seq_lengths_q`, which this
breaks.

The call now rebuilds the plan for the rows it is actually passing, with `cu_seqlens_q` clamped to match. A replayed full graph never reaches that branch, so the captured plan is left intact. Since the branch regenerates the plan anyway, it also takes the top-k lengths from the indices being passed rather than from the prediction made before the indexer ran - the sort leaves every `-1` at the tail of its row, so counting the non-negative entries gives exactly the left-aligned prefix the operator contract asks for (`sparse_flash_mla/README.md`, 约束说明), and unlike a prediction it cannot overshoot into the `-1` tail.

#### 4. Two masks were bounded by the padded token count, so neither ever fired

Both the top-k length mask in `SparseMLAMetadataState.prepare()` and the output mask in `sparse_mla()` compared against `query_start_loc[-1]`. The runner extends that terminator to `num_input_tokens` so the TND constraint holds,
which makes it equal to the tensor length being masked - the masks were no-ops.

The first one left padding rows claiming a top-k length derived from positions an earlier, larger batch had written, while their sequence lengths were zero. The second one was there to stop NaNs in unwritten graph-capacity rows from reaching the value and output projections, and never did. Both now use `num_actual_tokens`.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Manually validated GLM5Next A5 sparse MLA inference with the updated parameters.

Confirmed the previous sinks must be provided error no longer occurs.

test on gsm8k:
image

test on aime26:

image

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/attention/sparse_flash_mla.py [hunk 1]
```diff
@@ -60,6 +60,24 @@ def sparse_flash_mla_metadata(**kwargs):
...
+    CANN SparseFlashMla op still requires a per-head float32 sink tensor.
+    Keep this behavior here in the op adapter instead of leaking it into SFA
+    metadata construction.
+    """
+    if kwargs.get("sinks") is None:
+        kwargs["sinks"] = torch.ones(
+            (q.shape[1],),
+            dtype=torch.float32,
+            device=q.device,
+        )
```
[the remark this answers is not in the document]
Comment (zzzzzz198) [reply vllm_ascend/attention/sparse_flash_mla.py:78]: It‘s outdates now.

File: vllm_ascend/attention/sfa_v1.py [hunk 2]
```diff
@@ -72,6 +70,11 @@
 # NoPE sparse MLA operator helpers.
 SMLA_METADATA_SIZE = 1024
 SPARSE_ATTENTION_MAX_BLOCK_SIZE = 1024
+# GLM5Next's SFA path carries no learnable attention sink, but SparseFlashMla
+# still requires a per-head float32 sinks tensor. This is the placeholder value
+# the path has always used; whether -inf is the correct "no sink" value is a
+# separate question, tracked outside this change.
+SMLA_DEFAULT_SINK_VALUE = 1.0
```
Comment (lijiahang226) [inline vllm_ascend/attention/sfa_v1.py:78]: GLM5Next does not have an attention sink, but sinks=1.0 is not a neutral placeholder here. SparseFlashMla includes exp(sinks[h]) in the softmax denominator, so this changes the attention output numerically.

Comment (lijiahang226) [reply vllm_ascend/attention/sfa_v1.py:78]: Could you please provide some benchmark accuracy results, for example the score on AIME 2026, to help verify the correctness and accuracy impact of this change?

File: vllm_ascend/attention/sfa_v1.py [hunk 3]
```diff
@@ -124,9 +124,9 @@ def sparse_mla(query, cache, indices, metadata, scale):
             softmax_scale=scale,
             cmp_ratio=1,
             ori_mask_mode=3,
-            cmp_mask_mode=3,
-            ori_win_left=0,
-            ori_win_right=0,
+            cmp_mask_mode=0,
+            ori_win_left=-1,
+            ori_win_right=-1,
```
[the remark this answers is not in the document]
Comment (zzzzzz198) [reply vllm_ascend/attention/sfa_v1.py:214]: We also add it in build_smla_metadata now.

File: vllm_ascend/attention/sfa_v1.py (hunk 2 above)
Comment (zzzzzz198) [reply vllm_ascend/attention/sfa_v1.py:78]: We tested this PR on datasets, and for gsm8k the score is 97.12; for aime26 the score is 93.33. The results have been appended in PR description.

**Review by freyfwt — APPROVED** [review]

**Review by ningjingbengxiaohai — DISMISSED** [review]

**Comment by zzzzzz198** [comment]
/rerun
[Bot]: rerun completed. No failed jobs found.

File: vllm_ascend/attention/sparse_flash_mla.py [hunk 4]
```diff
@@ -70,5 +88,6 @@ def sparse_flash_mla(q: torch.Tensor, **kwargs):
         kwargs["seqused_ori_kv"] = kwargs.pop("seqused_kv")
     _drop_paged_kv_cu_seqlens(kwargs)
     _add_compressed_kv_lengths(kwargs)
+    _ensure_sinks(kwargs)
```
Comment (lcfenglinwan) [inline vllm_ascend/attention/sparse_flash_mla.py:91]: Is this an operator requirement or a model requirement? If it is a model requirement and is added here, it might affect other models.

Comment (zzzzzz198) [reply vllm_ascend/attention/sparse_flash_mla.py:91]: It is required by the operator. We have verified this with owner of the operator.

**Review by lcfenglinwan — APPROVED** [review]
LGTM

**Review by lijiahang226 — APPROVED** [review]

**Comment by zzzzzz198** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by ningjingbengxiaohai — APPROVED** [review]
